src/team_lib/mhba/xccl_mhba_ib.c

```diff
@@ -123,3 +124,213 @@ xccl_status_t xccl_mhba_qp_connect(struct ibv_qp *qp, uint32_t qp_num,
     }
     return XCCL_OK;
 }
+
+xccl_status_t xccl_mhba_init_dc_qps_and_connect(xccl_mhba_team_t *mhba_team, uint32_t *local_data, uint8_t port_num){
+    int i;
+    struct ibv_qp_init_attr_ex attr_ex;
+    struct mlx5dv_qp_init_attr attr_dv;
+    struct ibv_qp_attr qp_attr_to_init;
+    struct ibv_qp_attr qp_attr_to_rtr;
+    struct ibv_qp_attr qp_attr_to_rts;
+    memset(&attr_ex, 0, sizeof(attr_ex));
+    memset(&attr_dv, 0, sizeof(attr_dv));
+    memset(&qp_attr_to_init, 0, sizeof(qp_attr_to_init));
+    memset(&qp_attr_to_rtr, 0, sizeof(qp_attr_to_rtr));
+    memset(&qp_attr_to_rts, 0, sizeof(qp_attr_to_rts));
+
+    attr_ex.qp_type = IBV_QPT_DRIVER;
+    attr_ex.send_cq = mhba_team->net.cq;
+    attr_ex.recv_cq = mhba_team->net.cq;
+    attr_ex.pd = mhba_team->node.shared_pd;
+    attr_ex.cap.max_send_wr = (SQUARED(mhba_team->node.sbgp->group_size / 2) + 1) * MAX_OUTSTANDING_OPS *
+                              xccl_round_up(mhba_team->net.net_size, NUM_DCI_QPS);
+    attr_ex.cap.max_send_sge = 1;
+    attr_ex.comp_mask |= IBV_QP_INIT_ATTR_SEND_OPS_FLAGS | IBV_QP_INIT_ATTR_PD;
+    attr_ex.send_ops_flags = IBV_QP_EX_WITH_RDMA_WRITE | IBV_QP_EX_WITH_RDMA_WRITE_WITH_IMM |
+                             IBV_QP_EX_WITH_ATOMIC_FETCH_AND_ADD;
+    attr_dv.comp_mask |= MLX5DV_QP_INIT_ATTR_MASK_DC | MLX5DV_QP_INIT_ATTR_MASK_QP_CREATE_FLAGS;
+    attr_dv.dc_init_attr.dc_type = MLX5DV_DCTYPE_DCI;
+    attr_dv.create_flags |= MLX5DV_QP_CREATE_DISABLE_SCATTER_TO_CQE;
+
+    qp_attr_to_init.qp_state = IBV_QPS_INIT;
+    qp_attr_to_init.pkey_index = 0;
+    qp_attr_to_init.port_num = port_num;
+    qp_attr_to_init.qp_access_flags = IBV_ACCESS_LOCAL_WRITE |
+                                      IBV_ACCESS_REMOTE_READ |
+                                      IBV_ACCESS_REMOTE_WRITE | IBV_ACCESS_REMOTE_ATOMIC;
+
+    qp_attr_to_rtr.qp_state = IBV_QPS_RTR;
+    qp_attr_to_rtr.path_mtu = IBV_MTU_4096;
+    qp_attr_to_rtr.min_rnr_timer = 20;
+    qp_attr_to_rtr.ah_attr.port_num = port_num;
+    qp_attr_to_rtr.ah_attr.is_global = 0;
+
+    qp_attr_to_rts.qp_state = IBV_QPS_RTS;
+    qp_attr_to_rts.timeout = 10; //todo - what value?
+    qp_attr_to_rts.retry_cnt = 7;
+    qp_attr_to_rts.rnr_retry = 7;
+    qp_attr_to_rts.sq_psn = 0x123;
+    qp_attr_to_rts.max_rd_atomic = 1;
+
+    //create DCIs
+    for (i =0; i<NUM_DCI_QPS ;i++) {
```

Comment:
why NUM_DCI_QPS is compile time constant instead of dynamical parameter?

Reply:
because it's a fixed value we will have, probably 16. it is equivalent to the number of threads we will use in APU in the future. According to Rami's calculations, 16 should be the number to achieve full BW

Reply:
do you mean adding that as a run parameter, like Transpose?

Reply:
yes, that is what i meant - dynamic env var parameter